kaiax/gov/headergov/impl/consensus.go

```diff
@@ -0,0 +1,156 @@
+package impl
+
+import (
+	"reflect"
+
+	"github.com/kaiachain/kaia/blockchain/state"
+	"github.com/kaiachain/kaia/blockchain/types"
+	"github.com/kaiachain/kaia/blockchain/types/account"
+	"github.com/kaiachain/kaia/common"
+	"github.com/kaiachain/kaia/kaiax/gov"
+	"github.com/kaiachain/kaia/kaiax/gov/headergov"
+)
+
+func (h *headerGovModule) VerifyHeader(header *types.Header) error {
+	if header.Number.Uint64() == 0 {
+		return nil
+	}
+
+	// 1. Check Vote
+	if len(header.Vote) > 0 {
+		vote, err := headergov.DeserializeHeaderVote(header.Vote)
+		if err != nil {
+			logger.Error("Failed to parse vote", "num", header.Number.Uint64(), "err", err)
+			return err
+		}
+
+		err = h.VerifyVote(header.Number.Uint64(), vote)
+		if err != nil {
+			logger.Error("Failed to verify vote", "num", header.Number.Uint64(), "err", err)
+			return err
+		}
+	}
+
+	// 2. Check Governance
+	if header.Number.Uint64()%h.epoch != 0 {
+		if len(header.Governance) > 0 {
+			logger.Error("governance is not allowed in non-epoch block", "num", header.Number.Uint64())
+			return ErrGovInNonEpochBlock
+		} else {
+			return nil
+		}
+	}
+
+	expected := h.getExpectedGovernance(header.Number.Uint64())
+	if len(header.Governance) == 0 {
+		if len(expected.Items()) != 0 {
+			return ErrGovVerification
+		}
+
+		return nil
+	}
+
+	actual, err := headergov.DeserializeHeaderGov(header.Governance)
+	if err != nil {
+		logger.Error("Failed to parse governance", "num", header.Number.Uint64(), "err", err)
+		return err
+	}
+
+	if !reflect.DeepEqual(expected, actual) {
+		logger.Error("Governance mismatch", "expected", expected, "actual", actual)
+		return ErrGovVerification
+	}
+
+	return nil
+}
+
+func (h *headerGovModule) PrepareHeader(header *types.Header) error {
+	// if epoch block & vote exists in the last epoch, put Governance to header.
+	if len(h.myVotes) > 0 {
+		header.Vote, _ = h.myVotes[0].Serialize()
+	}
+
+	if header.Number.Uint64()%h.epoch == 0 {
+		gov := h.getExpectedGovernance(header.Number.Uint64())
+		if len(gov.Items()) > 0 {
+			header.Governance, _ = gov.Serialize()
+		}
+	}
+
+	return nil
+}
+
+func (h *headerGovModule) FinalizeHeader(header *types.Header, state *state.StateDB, txs []*types.Transaction, receipts []*types.Receipt) error {
+	return nil
+}
+
+// VerifyVote takes canonical VoteData and performs the semantic check.
+func (h *headerGovModule) VerifyVote(blockNum uint64, vote headergov.VoteData) error {
+	if vote == nil {
+		return ErrNilVote
+	}
+
+	// TODO: check if Voter is valid.
+	// TODO: check if Voter is the block proposer.
```

Comment:
If we don't check if Voter is the proposer, any GC can spoof the VoteData so voter can be the governing node.

> this semantic has been preserved til now implicitly in header governance.
To prevent attacks I think we should check it explicitly.

Reply:
I'm unclear on the context in which 'spoof' is being used here. Could you explain it further? Also, if the check confirms that the voter is the proposer at the time of the `Vote()` API call, wouldn't the check pass if they are the proposer for that block? I'm not quite following the intent behind your question

Reply:
A malicious node X can propagate a vote `{ Voter: V, Name: "", Value: 1}`, where V is the governing node, when in fact the Voter should be X. Forbidding it from API call is insufficient, because it can change the logic. So we verify it in consensus that when X is the proposer, the voter must also be X.

It's possible because `header.Vote` is parsed to `{Voter, Name, Value}`, each of which can be an arbitrary data.

Reply:
Thinking about your comment, actually this check should only be done from consensus, not from the vote API because the API should be able to handle the vote request anytime, not only when it's proposer of the current block. I should fix it (maybe separate `VerifyVote` for consensus and API), thanks.

Reply:
Ah, verifying voting integrity should have been included—it's extremely important. Thank you.